File: day_1/solution.hs
import qualified Data.Set as S (fromList)
import System.IO (withFile, IOMode( ReadMode ), hGetContents)

calculate :: (Integral a) => a -> a
calculate x
    | weight <= 0 = 0
    | otherwise   = weight + (calculate weight)
  where
      weight = (x `quot` 3) - 2

main = withFile
    "input_data" ReadMode (\handle -> do
        contents <- ((map calculate) . (map read) . lines) <$> hGetContents handle
        print $ (sum . (S.fromList)) contents)

File: day_1/solution.py
with open("input_data") as file:
    modules = [int(line.strip()) for line in file if line]

raw_fuel_requirement = sum({(weight // 3) - 2 for weight in modules})

# Part 2

def calculate(weight: int) -> int:
    if (amount := (weight // 3) - 2) <= 0:
        return 0

    return amount + calculate(amount)

fuel_requirement = sum({calculate(weight) for weight in modules})
